core-bundle/src/Resources/public/mootao.js

```diff
@@ -80,12 +82,23 @@ Request.Contao = new Class(
 	failure: function() {
 		var url = this.getHeader('X-Ajax-Location');
 
+		this.redirectIfUnauthorized();
+
 		if (url && this.options.followRedirects && this.status >= 300 && this.status < 400) {
 			location.replace(url);
 			return;
 		}
 
 		this.onFailure();
+	},
+
+	redirectIfUnauthorized: function() {
+		var url = this.getHeader('X-Ajax-Location'),
+			unauthorized = null !== this.getHeader('X-Is-Unauthorized');
```

Comment:
Why do we need a custom header? Can we not use the existing HTTP status codes?

Reply:
Maybe I was not clear enough. Instead of

```php
        if ($request->isXmlHttpRequest()) {
             return new Response($location, 303, ['X-Is-Unauthorized' => 'true', 'X-Ajax-Location' => $location]);
         }
```

we should do

```php
        if ($request->isXmlHttpRequest()) {
             return new Response('Unauthorized', 401, ['X-Ajax-Location' => $location]);
         }
```

in the AuthenticationEntryPoint class.

Reply:
Ah yes of course, updated!